### ISSUE TITLE
openvpn: Respect the radius Framed-IP-Address attribute for client ip assignment

### ISSUE THREAD

**Issue by EugenMayer** [opnsense/core#2076]
Yet, one can choose Radius as an authentication for openvpn already, but assigning an IP is yet not possible, since the Framed-IP-Address attribute is yet not respected.

https://wiki.freeradius.org/guide/Ippool-and-radius-clients should be a reference 

![image](https://user-images.githubusercontent.com/136934/34780756-3f8d058e-f624-11e7-81f9-e3fc346dcc83.png)

As far as i understand, one would do ccd for openvpn - at least that is how i was doing that in the past with custom bash foo when booking in a new client. Not sure there is a better way not involving creating files on the FS?

Problably something where we could neat-pick? https://roessner-network-solutions.com/openvpn-radius-mysqlldap-howto/

Offering funding of this one.

**Comment by mimugmail** [comment]
Just to add it here: https://forum.opnsense.org/index.php?topic=5883.0
Would love this too since this would allow "real" user management via API :)

**Comment by EugenMayer** [comment]
We had a chat with @fichtner about how to implement this, i try to rephrase / not down what was "decided" and hope i nail it - if not, punch me.


1. we create a ccd.php in https://github.com/opnsense/core/tree/master/src/etc/inc/plugins.inc.d/openvpn which handle the actuall resolving of ccd meta data. 
So in here, we will ask, who ever wants to provide, in this case its the freeradius server, what the meta-data for the client are. We will handle over the commono name from the auth, which already has been happened ( and succeded ).  After fetching the clients information from <radius> we will write them with https://github.com/opnsense/core/blob/master/src/etc/inc/plugins.inc.d/openvpn.inc#L992 as a ccd entry and return. If not client infos could be fetched, we should *DELETE* the ccd file in the openvpn server ccd folder for this folder

2. we call the ccd.php script from https://github.com/opnsense/core/blob/master/src/etc/inc/plugins.inc.d/openvpn/ovpn_auth_verify#L13 so right after the auth has been verified, passing along all the client details like common name. this will be called always right now, we do not have a "disable ccd" in the openvpn configuration. 

(If we introduce such an option, we would need to pass it on here https://github.com/opnsense/core/blob/master/src/etc/inc/plugins.inc.d/openvpn.inc#L684 )

**Comment by EugenMayer** [comment]
I started with the implementation. I added a checkbox to the openvpn server config "user dynamic ccd lookup" so one can enable and disable this for a openvpn server

A core issue now @fichtner is, that we must use the auth-response of the radius server, which includes the CCD payload ( attributes ). We cannot re-fetch those informations, since in OTP environments this would need a reauthentication.

digging in this "classy legacy code" ( one hack stacking on the other .. holy shit )

https://github.com/opnsense/core/blob/master/src/etc/inc/auth.inc#L787

This is at least a call into a proper API ( thank you fitch) but the response is only allowed to be a boolean, same goes for `authenticate_user` itself, so i cannot get a complex payload here https://github.com/opnsense/core/blob/master/src/etc/inc/plugins.inc.d/openvpn/auth-user.php#L115 .. not to speak about that i would need to pipe all that on stdout and pick it up from  stdin here https://github.com/opnsense/core/blob/master/src/etc/inc/plugins.inc.d/openvpn/ovpn_auth_verify#L11

This is way to complicated with the current authentication / procedural flow.

That said, there is no way to create a CCD "generic" API right now, i will need to put the CCD generation into the authentication code of freeradius ( even though it does not belong there ).

Said that, we are here now https://github.com/opnsense/core/blob/master/src/opnsense/mvc/app/library/OPNsense/Auth/Radius.php#L341 and since the only arguments we have is username and password, we have no clue what opnvpn server asked for authentication thus we cannot generate anything for "him".

----

Bottom line, no way for "generate CCD during auth" - we have to statically generate them cron based / event based.

**Comment by EugenMayer** [comment]
Ok looked at your plugin @mimugmail and happened to see that the whole radius configuration is file based and you generate the users on the file system.

So what i would do now is, extend your plugin, not only generating the users into the FreeRadius configuration, but also

1. check for all openvpn servers which have "dynamic ccd lookup" active
2. generate the ccd for them int their folder

Is that a go?

**Comment by AdSchellevis** [comment]
@EugenMayer I haven't read all of the issue/request, but the call flow for the auth framework is as followed:

1. AuthenticationFactory->get("server_name") , delivers an IAuthConnector type object
2. next call authenticate() to request if the user is authenticated
3. if the authenticator is able to send information back, such as accounting information, you can call getLastAuthProperties() to request it back from the authenticator.

So, if the data you need is send back during the authentication you should be able to stash it in the object for retrieval...

**Comment by mimugmail** [comment]
@EugenMayer hitting Apply or Save calls reconfigure which also reloads the Templates in Service Controller (every file gets generated). 

Technically it should work to cross generate model content, but unsure if this is political correct.

**Comment by EugenMayer** [comment]
Found +TARGETS which was basically my missing link here to the namespace based template trigger described in https://docs.opnsense.org/development/architecture.html#backend-architecture and the configuration.

@mimugmail as you mentioned, now i need to trigger a foreign template generation, which i do not even think we technically have here, since we use a different data stream.

Since openvpn has no mvc plugin, there are no models or complex things to call anyway, its fairly straight forward. I will try that. Thanks

**Comment by EugenMayer** [comment]
Implemented the plugin part now - finished and teste:

Plugin: https://github.com/opnsense/plugins/pull/485
Core: https://github.com/opnsense/core/pull/2084

You can test it using https://github.com/EugenMayer/opnsense-openvpn-radius

You will need the `xmlstarlet` and `base64` binary installed on your host - due to the config.xml manipulation.
You can avoid it by commenting this out https://github.com/EugenMayer/opnsense-openvpn-radius/blob/master/Vagrantfile#L76

Then
```
git clone https://github.com/EugenMayer/opnsense-openvpn-radius
make start
```

Login via "root/opnsense" and check the 2 openvpn servers and the freeradius server. CRUD on client side overrides / users in freeradius and check `/var/etc/openvpn-csc/`

`/var/etc/openvpn-csc/1` will change ( enabled dynamic ccd ) while `/var/etc/openvpn-csc/2` will not

**Comment by EugenMayer** [comment]
closing the effort since i wont continue. If anybody want to pick this one up, feel free to reopen the issue anytime

**Comment by mimugmail** [comment]
@AdSchellevis 
Second FR via Forums:
https://forum.opnsense.org/index.php?topic=7830.0

If you like to go on with your idea to add the Framed-IP-Address via Radius I'd be happy to test/help.

**Comment by AdSchellevis** [comment]
@mimugmail it's not at the top of my priority list at the moment, but if you create an issue, I'll try to collect the plan into it (parts are scattered over different pulls and issues now) and assign it to myself.
